File: 617-merge-two-binary-trees/617-merge-two-binary-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {TreeNode}
 */
var mergeTrees = function(root1, root2) {
    if(!root1) return root2;
    if(!root2) return root1;
    
    let st1 = [];
    let st2 = [];
    
    st1.push(root1);
    st2.push(root2);
    
    while(st1.length > 0 && st2.length > 0) {
        let cur1 = st1[st1.length-1];
        let cur2 = st2[st2.length-1];
        
        // console.log(cur1.val, cur2.val);
        st1.pop();
        st2.pop();
        
        if(cur1 && cur2) {
            cur1.val += cur2.val;
        }
        
        if(cur1.left && cur2.left) {
            st1.push(cur1.left);
            st2.push(cur2.left);
        } else if(cur2.left) {
            cur1.left = new TreeNode(0);
            st1.push(cur1.left);
            st2.push(cur2.left);
        } else if(cur1.left) {
            cur2.left = new TreeNode(0);
            st1.push(cur1.left);
            st2.push(cur2.left);
        }
        
        if(cur1.right && cur2.right) {
            st1.push(cur1.right);
            st2.push(cur2.right);
        } else if(cur2.right) {
            cur1.right = new TreeNode(0);
            st1.push(cur1.right);
            st2.push(cur2.right);
        } else if(cur1.right) {
            cur2.right = new TreeNode(0);
            st1.push(cur1.right);
            st2.push(cur2.right);
        }
    }
    
    return root1;
};

File: 617-merge-two-binary-trees/617-merge-two-binary-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* mergeTrees(TreeNode* root1, TreeNode* root2) {
        if(!root1) return root2;
        if(!root2) return root1;
        
        stack<pair<TreeNode*, TreeNode*>> st;
        stack<TreeNode*> mergedSt;
        TreeNode* mergedTree = new TreeNode(root1 -> val + root2 -> val);
        TreeNode* itr = mergedTree;
        
        st.push({root1, root2});
        mergedSt.push(itr);
        
        while(!st.empty()) {
            root1 = st.top().first;
            root2 = st.top().second;
            itr = mergedSt.top();
            st.pop();
            mergedSt.pop();
            
            if(root1 && root2 && root1 -> left && root2 -> left) {
                itr -> left = new TreeNode(root1 -> left -> val + root2 -> left -> val);
                st.push({root1 -> left, root2 -> left});
                mergedSt.push(itr -> left);
            } else if(root1 && root1 -> left) {
                itr -> left = new TreeNode(root1 -> left -> val);
                st.push({root1 -> left, NULL});
                mergedSt.push(itr -> left);
            } else if(root2 && root2 -> left) {
                itr -> left = new TreeNode(root2 -> left -> val);
                st.push({NULL, root2 -> left});
                mergedSt.push(itr -> left);
            }
            
            if(root1 && root2 && root1 -> right && root2 -> right) {
                itr -> right = new TreeNode(root1 -> right -> val + root2 -> right -> val);
                st.push({root1 -> right, root2 -> right});
                mergedSt.push(itr -> right);
            } else if(root1 && root1 -> right) {
                itr -> right = new TreeNode(root1 -> right -> val);
                st.push({root1 -> right, NULL});
                mergedSt.push(itr -> right);
            } else if(root2 && root2 -> right) {
                itr -> right = new TreeNode(root2 -> right -> val);
                st.push({NULL, root2 -> right});
                mergedSt.push(itr -> right);
            }
        }
        
        return mergedTree;
    }
};